### ISSUE TITLE
FastWAM Lab 01: every recipe is 0% closed-loop (fails before approach); pad7 + unfreeze-4 still 0/10

### ISSUE THREAD

**Issue by alexhegit** [rocPAI-Forge/so101-simstudio#8]
## Summary

Lab 01 FastWAM from [`lerobot/fastwam_base`](https://huggingface.co/lerobot/fastwam_base) on the **same 50-episode** joint-position dataset is a **closed joint-space LIBERO→SO-101 transfer failure**. It dies **before approach** (`reach_8cm=0`; cube stays at spawn `z≈0.015`). Same data: ACT 6-D **~58%**, MolmoAct2 6-D **~54%**.

Best measured recipe (LIBERO **pad7** + unfreeze last **4** Wan video MoT layers) is still **0/10** (also **0/5 GLFW** GUI). Inference is **not** the bottleneck (~14–36 Hz). **Not on Hub** (weights live on DORobot only). Classroom policy stays **ACT 6-D / MolmoAct2 6-D**.

**Ledger:** Lab 01 [§6.4](https://github.com/rocPAI-Forge/so101-simstudio/blob/main/labs/lab01_pnp/lab01_pnp.md) / [§6.4.2](https://github.com/rocPAI-Forge/so101-simstudio/blob/main/labs/lab01_pnp/lab01_pnp.md)

Siblings: VLA-JEPA (grasp-stage, Hub 10K); GR00T [issue #6](https://github.com/rocPAI-Forge/so101-simstudio/issues/6) (approach yes / lift no).

## Failure data (closed-loop)

`reset_arm: home`, sync. Cameras: `top→image`, `wrist→image2`. Cube never leaves spawn unless noted. On MI300X, EGL `EGL_BAD_ALLOC` → OSMesa.

| Recipe | Protocol | Backend | n | Success | Notes |
|--------|----------|---------|---|:-------:|-------|
| 4K 6-D, freeze video, `lambda_video=0`, reinit 6-D path | Full-range, keyboard `home` | OSMesa | 20 | **0/20** | Infer **33.5 Hz**; ckpt `.../004000/` |
| 20K resume from 4K, freeze video, bs4 | Full-range, keyboard `home` | OSMesa | 10 | **0/10** | Infer **35.8 Hz**; loss ~0.111; `.../020000/` |
| same 20K | Full-range, demo-mean `home_joints` | OSMesa | 10 | **0/10** | `n_action_steps=10`; infer **18.1 Hz** |
| LIBERO pad7 20K (inherit 7-D/8-D, pad 6 joints) | Full-range | OSMesa | 10 | **0/10** | Open-loop RMSE 0.198 rad (better, still 0%) |
| **pad7 + unfreeze last 4 MoT 8K** (best) | Full-range, keyboard `home` | OSMesa | 10 | **0/10** | RMSE **0.136** rad (~3.8× ACT); `reach_8cm=0` |
| same pad7 unfreeze-4 | Full-range | **GLFW GUI** (.43) | 5 | **0/5** | Visual confirm; cube `z≈0.0149`; arm sweeps **not toward cube**; infer **13.9 Hz** |

Matched n=3 audit: FastWAM **0/3** at `reach_8cm`. Debug dump: arm is **not** frozen (`shoulder_pan` range ~0.39 rad in first 80 ticks) — motion is just not task-aligned.

Does **not** pass start RMSE `<0.08` / cube–action `r>0.5`. Gate status **closed**.

DORobot paths (do not Hub without a non-zero `k/n`):  
`/mnt/doscratch/outputs/train/lab01_pnp_fastwam/`  
pad7 unfreeze-4: DORobot-only (runbook: not published).

## What worked (do not undo)

- 6-D `*.pos` + `LAB01_STATE_DIM=6`. Cross-embodiment loader **reinitializes** mismatched action/proprio tensors (`fastwam_base` is 7-D / 8-D).
- Pad7 inherits the LIBERO action pathway instead of a dead reinit — **open-loop improved**, closed-loop still 0%.
- Freeze ~5B Wan **video** expert + `lambda_video=0` was the VRAM-safe Lab 01 default (bs4 on MI300X).
- ROCm extras: `scripts/install-fastwam-deps.sh` / `install-lerobot-extra-deps.sh`. Never pull CUDA torch into `.venv-rocm`.
- GUI GLFW check on halo confirmed OSMesa 0% was not a software-render artifact.

## Failed recipes (do not repeat as first levers)

1. **Train action DiT on frozen Wan, reinit 6-D** — base **has no** matching action-pathway tensors; underfit (RMSE ~0.23 rad).
2. **Demo-mean start + shorter chunks** — still never leave spawn. Keyboard-home OOD is **not sufficient** to explain 0%.
3. **Unfreeze last 4 video MoT** — open-loop RMSE 0.198→0.136; **closed-loop still 0%**. Do not unfreeze the full 5B video expert as the next cheap lever.
4. **More steps on freeze-video 20K** — 0/10 after 4K already 0/20.
5. Treating Lab 01 `lambda_video=0` as a test of the [Fast-WAM paper](https://arxiv.org/abs/2603.16666). The paper’s claimed causal ingredient is **video co-training**. Lab 01 never enabled it; that is a **documented gap**, not a remaining classroom lever on 50 SO-101 episodes (JEPA already showed adding a marketed WM term here does not yield a classroom policy).

## Lessons

1. **Death is before approach** — unlike JEPA (grasp) and GR00T 10K (lift). Pixel sensitivity can exist while **robot-space grounding** is wrong (FK min ~12.8 cm in the unfreeze-4 probe).
2. Inference Hz was fine; **do not** spend the next cycle on faster kernels.
3. Open-loop RMSE moving toward ACT is **necessary but not sufficient**; FastWAM still fails the closed-loop gate.
4. Camera rename is second-order (cube pixels still drove some response). First-order is LIBERO→SO-101 action/vision transfer on 50 episodes.

## Next work (tracked here)

- [ ] Do **not** open Lab 01 long trains for classroom success on this 50-ep set (`lambda_video>0`, more MoT layers, more pad7 steps).
- [ ] If a future experiment tests **paper-level video co-training**, use larger or **same-embodiment** data and keep it **out of the classroom path**.
- [ ] Do not Hub FastWAM until a Lab 01 protocol run has non-zero `k/n`.
- [ ] Classroom demos stay ACT 6-D / MolmoAct2 6-D.

## Eval

```bash
LAB01_POLICY_PATH=./outputs/train/lab01_pnp_fastwam/checkpoints/020000/pretrained_model \
LAB01_EVAL_CONFIG=labs/lab01_pnp/configs/rollout_fastwam.yaml \
LAB01_N_ACTION_STEPS=32 LAB01_EVAL_EPISODES=10 \
LAB01_RENDER_WINDOW=false LAB01_MUJOCO_GL=osmesa \
  ./labs/lab01_pnp/eval.cmd
```

**Comment by sattyamjjain** [comment]
One statistics question on the ranking. At n = 10, 0/10 here and 1/10 for VLA-JEPA in #7 are not distinguishable: the 95% interval for 0/10 reaches about 28%, and for 1/10 about 40%. Would you consider a fixed 20 to 50 episode, fixed-seed protocol for the final table, plus the reach, grasp, lift funnel you used for GR00T in #6, so the failed recipes can be told apart? Also, was the 0.136 rad open-loop RMSE measured on training or held-out episodes?

**Comment by alexhegit** [comment]
Good catch on the binomial overlap.

**On ranking / n.** Yes — for a final comparison table we should treat Wilson intervals as first-class and not rank FastWAM 0/10 vs JEPA 1/10 as if those point estimates were separable. Lab 01 already records that explicitly in §6.4.1 (0/10 → 0–28%, 1/10 → 2–40%). Classroom / reference rows (ACT, MolmoAct2, SmolVLA) already use fixed full-range **n=50**; the transfer-failure rows were left at n=10 / fixed-pose because the stop decision was mechanism-driven, not a claim that 0% vs 10% is statistically resolved.

What already separates the failed recipes without waiting on a 50-ep retake is the **stage funnel**, same family as GR00T #6:

| Recipe | Reach ≤8 cm | Near ≤4 cm | Close | Lift | Final |
|--------|:-----------:|:----------:|:-----:|:----:|:-----:|
| FastWAM pad7 unfreeze-4 (matched n=3) | **0/3** | 0 | 0 | 0 | 0 |
| JEPA BC 10K (matched n=3) | 3/3 | 3/3 | 3/3 | 3/3 | **0/3** (drop before box) |
| JEPA unfreeze-Qwen (matched n=3) | 3/3 | 3/3 | **0/3** | 0 | 0 |
| GR00T 10K (#6, n=10) | ~9–10/10 | … | … | **0/10** | 0 |

So FastWAM vs JEPA is distinguishable by **where** the policy dies (before approach vs after grasp), even when success rates both sit inside overlapping CIs. A fixed-seed **n=20–50** protocol on the failed rows would still be useful for the published table; it is not required to reopen Lab 01 long trains on this 50-ep set (§6.4.2 stop boundary).

**On the 0.136 rad RMSE.** It is **in-sample / training-episode**, not held-out. Every Lab 01 checkpoint here was trained on all 50 episodes, so `probe_open_loop.py` is an explicit teacher-forced diagnostic (start / 25% / 50% / 75% phases, deterministic seed). From §6.4.1: *"no honest held-out action test exists without retraining."* Canonical FastWAM unfreeze-4 row in that table is overall RMSE **0.127** rad (start **0.193**); the **0.136** cited in the issue / §6.4.2 ledger is the same in-sample family (earlier / rounded report of that probe), not a validation split. Gate status stays closed: neither number passes start RMSE `<0.08`.

**Comment by alexhegit** [comment]
A bit more context on *why* the near-zero closed-loop rates show up here, beyond the n=10 vs n=10 ranking question.

Two factors are doing most of the work, and they interact:

1. **Sparse coverage + noisy teleop on a large spawn box.** The 50-episode set still undersamples a fairly large random cube workspace. Trajectories were collected by a human teleoperating *outside* the sim (viewpoint mismatch, latency), so quality is uneven: jitter, slow/dragging motion, and re-grasps are common. That is enough to keep a policy from locking onto a clean approach even when the task is learnable (ACT / MolmoAct2 on the same set). We are now using **lab01-agent** to auto-expand a larger pool of more stable, higher-quality trajectories.

2. **Cross-embodiment / sensor mismatch on top of that data.** FastWAM, VLA-JEPA, LingBot-VA, and similar bases were pretrained with different camera counts and placements, often a **7-DoF** arm (SO-101 is **6-DoF**), and in some cases bimanual layouts. Mapping those into Lab 01’s 6-D joint-position, 1-arm, 2-cam setup adds another error term. The mix is messy: FastWAM dies before approach, JEPA around grasp/place, LingBot-VA is an EEF-language mismatch rather than a measured 0%. We are not treating those as one failure mode.

**Near-term plan:** put the effort on (1) — more and cleaner demos — then re-run a full eval pass on the expanded set. That is the intended next measurement, not more Lab 01 long trains on the current 50 episodes.

**Comment by sattyamjjain** [comment]
The stage funnel is the right answer, and it is better than the one I was angling for. A funnel row tells you where the recipe died; a scalar at n=10 only tells you it died. Reach, near, close, lift, final is enough resolution to separate "never approached" from "approached and could not close", and those two failures have different fixes.

One thing to guard while you use it. The funnel stages are nested, so the counts are not independent, and it is tempting to put an interval on each stage as if they were. If you want intervals on the funnel, the honest ones are conditional: close given near, lift given close. Reporting an unconditional interval per stage makes the later stages look more certain than they are.

On the two explanations for the near-zero closed-loop rates, sparse noisy teleop coverage and cross-embodiment mismatch: those are separable with an ablation you already have the data for. Score the same policy on the subset of spawn positions that are densely covered in the training set, and on the sparse tail. If the gap is mostly coverage, the dense subset recovers, and the tail does not. If it is embodiment, both stay flat. That is one extra evaluation pass over episodes you already ran, no new collection.

Auto-expanding the trajectory pool with lab01-agent will help the coverage half either way, but it will not tell you which half you fixed unless you take the before reading now.
